server-helpers.js

```diff
@@ -0,0 +1,171 @@
+  }
+
+  static authenticateUser = (userAccount, attemptedPassword, response) => {
+    try {
+      if (userAccount[0].password === attemptedPassword) {
+        return response.status(200).json(this.makeSecureUserResponse(userAccount[0]))
+      } else {
+        return response.status(422).json('An incorrect password was provided!')
+      }
+    } catch (error) {
+      console.error(error.message)
+    }
+  }
+
+  static findAuthorByNameOrEmail = async (queryType, queryValue) => {
+    return knex('authors')
+      .groupBy('id')
+      .select()
+      .having(queryType, '=', queryValue)
+  }
+
+  static userIsTaken = async (author) => {
+    const { name, email } = author 
+    const checkAgainst = ["name", "email"]
+    const result = [name, email].reduce(async (isTaken, thing, i) => {
+      const resolvedUsers = await Promise.resolve(isTaken)
+      return await this.findAuthorByNameOrEmail(checkAgainst[i], thing).then(user => {
+        return user.length > 0 ? resolvedUsers.concat(checkAgainst[i]) : resolvedUsers
+      })
+    }, [])
+    return await result
+  }
+}
+
+class PromptHelper {
+  static promptGenerator = (genre) => {
+    try {
+      const allPrompts = () => {
+        return knex("prompts").groupBy("id").select();
+      }
+      const filterPrompts = () => {
+        return genre === 'any' ? allPrompts() : allPrompts().having("genre", "=", genre);
+      }
+      return filterPrompts().then((prompts) => {
+        const randomIndex = Math.round(Math.random() * prompts.length - 1);
+        return prompts[randomIndex]
+      });
+    } catch (error) {
+      console.error(error.message);
+    }
+  }
+
+  static findSpecificPrompt = (id) => {
+    return knex("prompts")
+      .groupBy("id")
+      .select()
+      .having("id", "=", id)
+  }
+}
+
+class StoryHelper {
+  static checkForProblems = (story) => {
+    const requiredFields = ['contributions', 'contributors']
+    const problems = []
+    const missingFields = requiredFields.filter(field => !Object.keys(story).includes(field))
+    missingFields.length > 0 && problems.push(`Your entry is missing this data: ${missingFields.join(", ")}.`)
+    Object.keys(story).forEach(field => {
+      if (field === "contributions" && typeof story.contributions !== "string") {
+        problems.push('Contributions should be a string.')
+      } else if (field === 'contributors' && isNaN(story.contributors)) {
+        problems.push('Contributors should be denoted by their id, a number.')
+      } else if (field === 'prompt' || field === 'title' || field === 'id') {
+        problems.push(`${fields}s are not allowed to be updated`)
+      } else if (field === 'is_complete' && typeof story.is_complete !== "boolean") {
+        problems.push('Stories being complete should be true or false.')
+      } else if (field === 'created_at' || field === 'updated_at') {
+        problems.push('You can not manually update publishing dates')
+      }
+    })
+    return problems
+  }
```

Comment:
I really appreciate how thorough you got with this, this'll save us a lot in the long run